analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs

```diff
@@ -173,105 +170,106 @@ IEnumerable<TMessage> ReadProtobuf()
             }
-
             using var workspace = MSBuildWorkspace.Create();
             workspace.WorkspaceFailed += (_, failure) => Console.WriteLine(failure.Diagnostic);
 
             // Copy razor project directory and test case files to a temporary build location
             var tempPath = Path.Combine(Path.GetTempPath(), $"ut-razor-{Guid.NewGuid()}");
+            Directory.CreateDirectory(tempPath);
             try
             {
-                Directory.CreateDirectory(tempPath);
-
-                List<string> languages = new();
-                if (builder.ParseOptions != null && builder.ParseOptions.Any())
-                {
-                    foreach (var parseOption in builder.ParseOptions)
-                    {
-                        if (parseOption is CSharpParseOptions csharpParseOptions)
-                        {
-                            languages.Add(GetLanguageVersionReference(csharpParseOptions));
-                        }
-                    }
-                }
-                else
+                foreach (var langVersion in builder.ParseOptions.Cast<CSharpParseOptions>().Select(LanguageVersionReference).ToArray())
                 {
-                    languages.Add("latest");
+                    var projectRoot = Path.Combine(tempPath, langVersion);
+                    Directory.CreateDirectory(projectRoot);
+                    var csProjPath = PrepareRazorProject(projectRoot, langVersion);
+                    var razorFiles = PrepareRazorFiles(projectRoot);
+                    yield return new(workspace.OpenProjectAsync(csProjPath).Result.GetCompilationAsync().Result, razorFiles.ToArray());
                 }
+            }
+            finally
+            {
+                Directory.Delete(tempPath, true);
+            }
+        }
 
-                foreach (var lang in languages)
-                {
-                    Directory.CreateDirectory(Path.Combine(tempPath, lang));
-                    // Copy all the files
-                    // To improve: Paths are currently relative to entry assembly => needs to be duplicated in different projects for now.
-                    foreach (var file in Directory.GetFiles(@"TestCases\Razor\EmptyProject").Concat(builder.Paths.Select(TestCasePath)))
-                    {
-                        File.Copy(file, Path.Combine(tempPath, lang, Path.GetFileName(file)));
-                    }
-                    foreach (var snippet in builder.Snippets)
-                    {
-                        File.WriteAllText(Path.Combine(tempPath, lang, snippet.FileName), snippet.Content);
-                    }
-                    var csprojPath = Path.Combine(tempPath, lang, "EmptyProject.csproj");
-                    var destinationXml = XElement.Load(csprojPath);
+        private static string LanguageVersionReference(CSharpParseOptions options) =>
```

Comment:
This method only applies for razor compilation, but the name is a bit generic and it's not obvious if you need version > C#10 _always_ or only for _razor_.

Please move it as a local function inside `CompileRazor`, to signal the scope to the reader.
I would also probably rename it to just `Language` if it's local.